atsign-foundation/noports
# test: e2e tests iteration 2

**- What I did**
- test: new e2e test harness. Design goals
    1. easy and fast to run locally
    2. can be run via github actions
    3. support multiple types of 'executables' (e.g. dart binaries, c binaries, docker containers)
    4. easily configurable matrix of daemon versions, client versions and tests to execute
- fix: allow sshnp >= 5.1.x client to be used in Linux scripting environments when using the -x flag

**- How I did it**
In tests/e2e_all:
- Entry point: scripts/main.sh, which then runs the following:
  - setup_binaries.sh which downloads artifacts and compiles local branch binaries
  - start_daemons.sh
  - run_tests.sh
  - stop_daemons.sh
  - cleanup_tmp_files.sh
- All the individual scripts use shared functions which are in tests/e2e_all/scripts/common/*.include.sh

in packages/dart/sshnoports/bin/sshnp.dart
- put try/catch around getting the original lineMode and echoMode from stdin (because the Dart VM barfs on Linux if the stdin isn't an actual terminal)
- and capture the error for later
- and throw the error if the terminal is actually required (which it won't be in a scripting environment)

**- How to verify it**
- e2e_all job runs, and tests pass
- run locally (MacOS or Ubuntu). See tests/e2e_all/README.md for instructions (tl;dr : execute tests/e2e_all/scripts/main.sh), as well as a description of how this test rig works.

**Additional context**
Shout-out and endless thanks to @JeremyTubongbanua who built the first iteration of the NoPorts e2e test rig
